Extensions for GameObject, Transform, and String to provide convience methods

searica opened the pull request:
I had originally intended to break these into a few smaller PRs but since the goal is a release this weekend I merged them into a single PR. In general each commit corresponds to a set of related extension methods for each class and explains the rationale behind adding those methods, so that it's hopefully easier to accept/reject whatever subset of extensions are judged to be appropriate for Jotunn.

MSchmoecker commented on JotunnLib/Extensions/GameObjectExtension.cs:
```diff
@@ -104,6 +106,174 @@ public static GameObject OrNull(this GameObject @this)
+        /// <returns></returns>
+        public static bool HasComponent<T>(this GameObject gameObject) where T : Component
+        {
+            return gameObject.GetComponent<T>() != null;
```
Unity actually discourages doing null checks against Unity objects (https://docs.unity3d.com/Manual/BestPracticeUnderstandingPerformanceInUnity7.html at Null comparisons against UnityEngine.Object subclasses)

Thus, the method should be simplefied to `return gameObject.GetComponent<T>()`, at which point the whole method is rather redundent and doesn't have a purpose anymore. It should be removed in my opinion, same for the string overload.

searica replied:
Oh, that's very good to know. Thank you! Will remove that then.

searica replied:
And also update some of my own code in other projects now that I know that

MSchmoecker commented on JotunnLib/Extensions/GameObjectExtension.cs:
```diff
@@ -104,6 +106,174 @@ public static GameObject OrNull(this GameObject @this)
+        {
+            foreach (var compo in components)
+            {
+                if (gameObject.GetComponent(compo) != null)
```
```suggestion
                if (gameObject.GetComponent(compo))
```

because of the mentioned null checks. This also applies to the follwing methods.

MSchmoecker commented on JotunnLib/Extensions/GameObjectExtension.cs:
```diff
@@ -104,6 +106,174 @@ public static GameObject OrNull(this GameObject @this)
+            return true;
+        }
+
+        /// <summary>
+        ///     Check if GameObject or any of it's children
+        ///     have any of the specified components.
+        /// </summary>
+        /// <param name="gameObject"></param>
+        /// <param name="components"></param>
+        /// <returns></returns>
```
The includeInactive paramter is missing in the docs

searica replied:
Thanks for catching that, will fix that now!

MSchmoecker commented on JotunnLib/Extensions/GameObjectExtension.cs:
```diff
@@ -104,6 +106,174 @@ public static GameObject OrNull(this GameObject @this)
+        /// <param name="name"></param>
+        /// <param name="includeInactive">Whether to include inactive child objects in the search or not.</param>
+        /// <returns></returns>
+        public static T GetComponentInChildrenByName<T>(
```
I'm not sure what for the use case for this method is, why is the component serched by T and checked with a name?

searica replied:
I made it because there are a few cases in MoreVanillaBuildPrefabs where I need to search for a Component in a specific child object but I may not know what the transform tree looks like.

 I think I wrote it before I figured out how to search for a specific nested child or that Utils.FindChild existed. I'll update the PR to remove this one since it's better accomplished by using Utils.FindChild(name).GetComponent<T>() instead.

MSchmoecker commented on JotunnLib/Extensions/GameObjectExtension.cs:
```diff
@@ -104,6 +106,174 @@ public static GameObject OrNull(this GameObject @this)
+        /// <param name="childName">Name of the child object to search for.</param>
+        /// <param name="breadthFirst"> Whether to preform a breadth first or depth first search. Default is breadth first.</param>
+        /// <returns></returns>
+        public static Transform FindDeepChild(this GameObject gameObject, string childName, bool breadthFirst = true)
+        {
+            return gameObject?.transform.FindDeepChild(childName, breadthFirst);
+        }
```
Valheim already has the method `Utils.FindChild`, which does the same. Although Wrapping this into an extension method could still be done, it should call the vanilla method.

Also, the null conditional (.?) on the gameobect doesn't work as expedted because of the Unity lifecyle [(it can return true despite the object being destroyed)](https://stackoverflow.com/questions/62678228/why-does-c-sharp-null-conditional-operator-not-work-with-unity-serializable-vari) and should be removed anyway. It's not the purose if this method to test if the gameObject is null.

searica replied:
Thanks for letting me know, will update accordingly and also very much appreciate the link to help learn more about Unity (will make sure to account for that in my own code now).

MSchmoecker commented on JotunnLib/Extensions/StringExtensions.cs:
```diff
@@ -0,0 +1,121 @@
+                return char.ToUpper(s[0]) + s.Substring(1);
+        }
+
+        /// <summary>
+        ///     Returns an Empty string if value is null
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        public static string EmptyIfNull(this object value)
+        {
+            if (value == null)
+            {
+                return string.Empty;
+            }
+            return value.ToString();
+        }
```
This method is redundent and can be written as `string ?? string.Empty` or `string ?? ""`

searica replied:
Yep, I wrote that before I learned about the ?? operator in C#. I'll remove this one as well.

MSchmoecker commented on JotunnLib/Extensions/TransformExtensions.cs:
```diff
@@ -0,0 +1,59 @@
+        /// <param name="childName">Name of the child object to search for.</param>
+        /// <param name="breadthFirst"> Whether to preform a breadth first or depth first search. Default is breadth first.</param>
+        /// <returns></returns>
+        public static Transform FindDeepChild(
```
As mentioned above, this method is already implemented in `Utils.FindChild` and should call the vanilla method

searica commented:
I've pushed a new set of changes that hopefully incorporate the feedback you gave. I'm not sure that the way I rewrote HasAllComponents to avoid null checks is the best way though. 

Edit: I did some more reading on MonoBehavior and how Unity handles implicitly casting them to a bool, so I removed the counter in HasAllComponents to optimize it a bit more.

MSchmoecker reviewed (approved):
Looks good, thanks for the extensions
